src/compiler/visitorPublic.ts

```diff
@@ -22,35 +22,61 @@ import {
+ *   function _must_ be provided, and that function must be a type predicate.
+ *
  * @param node The Node to visit.
  * @param visitor The callback used to visit the Node.
  * @param test A callback to execute to verify the Node is valid.
  * @param lift An optional callback to execute to lift a NodeArray into a valid Node.
  */
-export function visitNode<T extends Node>(node: T, visitor: Visitor | undefined, test?: (node: Node) => boolean, lift?: (node: readonly Node[]) => T): T;
-
+export function visitNode<TIn extends Node | undefined, TVisited extends Node | undefined, TOut extends Node>(
+    node: TIn,
+    visitor: Visitor<NonNullable<TIn>, TVisited> | undefined,
+    test: (node: Node) => node is TOut,
+    lift?: (node: readonly Node[]) => Node,
+): TOut | (TIn & undefined) | (TVisited & undefined);
 /**
  * Visits a Node using the supplied visitor, possibly returning a new Node in its place.
  *
+ * - If the input node is undefined, then the output is undefined.
+ * - If the visitor returns undefined, then the output is undefined.
+ * - If the output node is not undefined, then it will satisfy the test function.
+ * - In order to obtain a return type that is more specific than `Node`, a test
+ *   function _must_ be provided, and that function must be a type predicate.
+ *
  * @param node The Node to visit.
  * @param visitor The callback used to visit the Node.
  * @param test A callback to execute to verify the Node is valid.
  * @param lift An optional callback to execute to lift a NodeArray into a valid Node.
  */
-export function visitNode<T extends Node>(node: T | undefined, visitor: Visitor | undefined, test?: (node: Node) => boolean, lift?: (node: readonly Node[]) => T): T | undefined;
-
-export function visitNode<T extends Node>(node: T | undefined, visitor: Visitor | undefined, test?: (node: Node) => boolean, lift?: (node: readonly Node[]) => T): T | undefined {
-    if (node === undefined || visitor === undefined) {
+export function visitNode<TIn extends Node | undefined, TVisited extends Node | undefined>(
+    node: TIn,
+    visitor: Visitor<NonNullable<TIn>, TVisited> | undefined,
+    test?: (node: Node) => boolean,
+    lift?: (node: readonly Node[]) => Node,
+): Node | (TIn & undefined) | (TVisited & undefined);
+export function visitNode(
+    node: Node,
+    visitor: Visitor | undefined,
+    test?: (node: Node) => boolean,
+    lift?: (node: readonly Node[]) => Node,
+): Node | undefined {
+    if (node === undefined) {
+            // If the input type is undefined, then the output type can be undefined.
         return node;
     }
 
-    const visited = visitor(node);
-    if (visited === node) {
-        return node;
-    }
+    const visited = visitor ? visitor(node) : node;
 
     let visitedNode: Node | undefined;
     if (visited === undefined) {
+        // If the visited node is undefined, then the visitor must have returned undefined,

```

Comment:
I wonder how often this happens in practice. Ideally, a call to `visitNode` with a `Node` (and not `undefined`) should always produce a `Node`, though the `visitor` we pass has a chance of returning `undefined` because we often share the same `visitor` function. We really shouldn't elide nodes during a single-node visit, only when visiting lists. 

If we are able to verify that this doesn't happen in practice in our code base, I'd be tempted to change this check to one that throws instead of returning `undefined`, so that we can remove all of the additional `Debug.checkDefined` calls later on.

Reply:
I added a `Debug.fail()` to this branch, 693 tests fail. It seems like a lot of them come from `visitImportClause` in `compiler/transformers/ts.ts`.

Reply:
I was misled given my terminal has a limit; changing that one only fixes about 10 tests. It's possible to "fix" some of them by avoiding calling `visitNode` but it seems weird.

Reply:
Given how much this hits our codebase, I'm not sure that this one's safe to do. One immediate counterexample I can think of for a single node visit to return undefined is when removing a token, e.g. if I'm transforming an optional parameter or property to a required one.